File: RpgSystem/CChrEffect.cpp
/** 
 *  AKARRA SOURCE CODE (c) Jens Bergensten
 *  See EULA.txt for information about this file.
 *
 *  File description (may not be accurate):
 *    A character effect is a spell effect which has been placed on an entity
 *    and needs to be updated every second.
 **/

#include "CChrEffect.h"
#include "IAttributeObject.h"
#include "CEffect.h"
#include "../PlayerSystem/AkarraCharacter.h"
#include "../library/basic/ccharstring.h"
#include "../library/FileSystem/CFileStream.h"

//#ifndef NULL
//#define NULL ((void*)0)
//#endif

namespace RpgSystem
{

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CChrEffect::CChrEffect()
{
	m_attackergid = -1;
	mp_effect = NULL;
	ma_effectname[0] = 0;
	m_life = 0;
	m_seconds = 0;
	m_skillsource = -1;
	m_identifier = 0;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CChrEffect::~CChrEffect()
{
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// This function is only used for player characters (not monsters). The version is the
// CAkarraCharacter file version.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
bool CChrEffect::read(FileSystem::CFileStream& file, int version)
{
	if (!file.isOk())
	{
		return false;
	}

	file.read(ma_effectname, sizeof(char), 32);
	file.read(&m_life, sizeof(int), 1);
	file.read(&m_seconds, sizeof(int), 1);
	if (version > Player::CHRFV_030301)
	{
		file.read(&m_identifier, sizeof(int), 1);
		file.read(&m_skillsource, sizeof(int), 1);
		file.read(&m_sourceCR, sizeof(float), 1);
	}

	m_attackergid = -1;
	mp_effect = NULL;

	return true;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// This function is only used for player characters (not monsters).
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
bool CChrEffect::write(FileSystem::CFileStream& file)
{
	if (!file.isOk())
	{
		return false;
	}

	file.write(ma_effectname, sizeof(char), 32);
	file.write(&m_life, sizeof(int), 1);
	file.write(&m_seconds, sizeof(int), 1);
	file.write(&m_identifier, sizeof(int), 1);
	file.write(&m_skillsource, sizeof(int), 1);
	file.write(&m_sourceCR, sizeof(float), 1);

	// note: no need to save the gid since it's illegal anyway when we reload

	return true;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
int CChrEffect::getID()
{
	return m_identifier;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// This function is used to test for stacking abilities for the effects. Effects placed on
// players by monsters, and certain stackable skills, return false and usually have their
// skill id numbers set to -1. Otherwise this function only returns true if the submitted
// id is the same as the current effect. If the function returns true it means that this
// effect should be removed.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
bool CChrEffect::isOverruled(int skill)
{
	return ((m_skillsource >= 0) && (m_skillsource == skill));
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// This function returns true if the effect that is associated with the runable affects
// the character's statistics, such as attack rating and resistances. The point with the
// function is to detect if the server needs to recalculate the character's stats when a
// runable is added or removed.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
bool CChrEffect::affectsStats()
{
	if (!mp_effect)
	{
		return false;
	}

	switch (mp_effect->u_type)
	{
		case EMT_BOOST:
		case EMT_WEAKEN:
		case EMT_BOOSTPERCENT:
		case EMT_WEAKENPERCENT:
		case EMT_IMPROVEASPEED:
		case EMT_REDUCEASPEED:
		case EMT_SPEEDPERCENT:
			return true;
	}

	return false;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
CEffectType* CChrEffect::getEffect()
{
	return mp_effect;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// Since the effects are modified depending on the caster's cast rating, we must remember
// and use the correct value when we calculate the importance of the effect.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
float CChrEffect::getCastRating()
{
	return m_sourceCR;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
char* CChrEffect::getEffectName()
{
	return ma_effectname;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CChrEffect::setEffect(CEffectType* effect)
{
	if (!effect)
	{
		return;
	}

	mp_effect = effect;

	// copy the effect's name
	strcpy(ma_effectname, effect->ua_dataname);
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
void CChrEffect::startEffect(CEffectType* effect, IAttributeObject* source, int identifier, int skill)
{
	m_identifier = identifier;
	m_skillsource = skill;
	
	if (!effect)
	{
		return;
	}
	setEffect(effect);

	// start the timer, if this effect has a duration value
	m_life = effect->getDuration(0, source);
	// since we count in milliseconds, we need to increase the value
	m_life *= 1000;

	m_attackergid = -1;
	m_sourceCR = 10.0f;
	if (source)
	{
		m_attackergid = source->getIdNumber();
		m_sourceCR = source->getAbility(ABILITY_CAST_RATING);
	}
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// This function decreases the effect's timer and checks if it's time to perform any kind
// of action. It returns false when it should be removed from the character.
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
bool CChrEffect::runEffect(IAttributeObject* target, int difftime)
{
	if (!target || !mp_effect)
	{
		return false;
	}

	// decrease timer
	m_life -= difftime;

	if (m_life <= 0)
	{
		return false;
	}

	m_seconds += difftime;

	// perform actions
	switch (mp_effect->u_type)
	{
		case EMT_POISON:
			if (m_seconds >= 4000)	// poison deal damage every 4 seconds
			{
				m_seconds -= 4000;
				target->harmAttributeByEffect(mp_effect->getParameter(0, m_sourceCR) * 4.0f, 0, m_attackergid, mp_effect);
			}			
			break;

		case EMT_LIGHT:
			break;

		case EMT_REGENERATE:
			if (m_seconds >= 4000)
			{
				m_seconds -= 4000;
				target->restoreAttributeByEffect(mp_effect->getParameter(1, m_sourceCR) * 4.0f, (int)mp_effect->getParameter(0), m_attackergid, mp_effect);
			}			
			break;
	}

	return true;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
int CChrEffect::getSourceID()
{
	return m_attackergid;
}

// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
// ---------- ---------- ---------- ---------- ---------- ---------- ---------- ----------
int CChrEffect::getRemainingLifeMillis()
{
	return m_life;
}

}

File: Src/RPG-System/Effects/CChrEffect.cs
using System;
using System.IO;
using RpgSystem.Effects;
using RpgSystem.Entities;

namespace RpgSystem.Effects
{
    /// <summary>
    /// A character effect is a spell effect which has been placed on an entity
    /// and needs to be updated regularly. Effects can be temporary buffs, debuffs,
    /// regeneration, poison, etc.
    /// </summary>
    public class CChrEffect
    {
        // Fields
        private int _attackerId;       // ID of the entity that caused this effect
        private EffectType _effect;    // The effect definition
        private string _effectName;    // Name of the effect (for serialization)
        private int _lifeTime;         // Remaining time in milliseconds
        private int _seconds;          // Time counter in milliseconds (renamed to match C++ code)
        private int _skillSource;      // Source skill ID that created this effect
        private float _sourceCastRating; // Cast rating when effect was applied
        private int _identifier;       // Unique identifier for network transmission

        // Properties
        public int Id => _identifier;
        public string EffectName => _effectName;
        public float CastRating => _sourceCastRating;
        public int SourceId => _attackerId;
        public int RemainingLifeMillis => _lifeTime;
        public EffectType Effect => _effect;

        /// <summary>
        /// Creates a new character effect
        /// </summary>
        public CChrEffect()
        {
            _attackerId = -1;
            _effect = null;
            _effectName = string.Empty;
            _lifeTime = 0;
            _seconds = 0;
            _skillSource = -1;
            _sourceCastRating = 0;
            _identifier = 0;
        }

        /// <summary>
        /// Reads effect data from a binary stream (used for player characters)
        /// </summary>
        public bool Read(BinaryReader reader, int version)
        {
            try
            {
                // Read effect name (32 chars)
                _effectName = new string(reader.ReadChars(32)).TrimEnd('\0');
                
                // Read timers
                _lifeTime = reader.ReadInt32();
                _seconds = reader.ReadInt32();
                
                // Read other data if version supports it
                if (version > PlayerFileVersions.V_030301)
                {
                    _identifier = reader.ReadInt32();
                    _skillSource = reader.ReadInt32();
                    _sourceCastRating = reader.ReadSingle();
                }
                
                // These values are not saved (entity IDs are not persistent)
                _attackerId = -1;
                _effect = null;
                
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// Writes effect data to a binary stream (used for player characters)
        /// </summary>
        public bool Write(BinaryWriter writer)
        {
            try
            {
                // Write effect name (32 chars)
                char[] nameChars = new char[32];
                if (!string.IsNullOrEmpty(_effectName))
                {
                    for (int i = 0; i < Math.Min(_effectName.Length, 32); i++)
                    {
                        nameChars[i] = _effectName[i];
                    }
                }
                writer.Write(nameChars);
                
                // Write timers
                writer.Write(_lifeTime);
                writer.Write(_seconds);
                
                // Write other data
                writer.Write(_identifier);
                writer.Write(_skillSource);
                writer.Write(_sourceCastRating);
                
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// Gets the ID of this effect
        /// </summary>
        public int GetID()
        {
            return _identifier;
        }

        /// <summary>
        /// Checks if this effect is overruled by a new effect with the same skill ID
        /// </summary>
        /// <returns>True if this effect should be removed</returns>
        public bool IsOverruled(int skill)
        {
            return (_skillSource >= 0) && (_skillSource == skill);
        }

        /// <summary>
        /// Checks if this effect modifies character stats
        /// </summary>
        public bool AffectsStats()
        {
            if (_effect == null)
                return false;

            switch (_effect.Type)
            {
                case (int)EffectMechanismType.Boost:
                case (int)EffectMechanismType.Weaken:
                case (int)EffectMechanismType.BoostPercent:
                case (int)EffectMechanismType.WeakenPercent:
                case (int)EffectMechanismType.ImproveAttackSpeed:
                case (int)EffectMechanismType.ReduceAttackSpeed:
                case (int)EffectMechanismType.SpeedPercent:
                    return true;
                default:
                    return false;
            }
        }

        /// <summary>
        /// Gets the effect type
        /// </summary>
        public EffectType GetEffect()
        {
            return _effect;
        }

        /// <summary>
        /// Gets the cast rating that was used when the effect started
        /// </summary>
        public float GetCastRating()
        {
            return _sourceCastRating;
        }

        /// <summary>
        /// Gets the effect name
        /// </summary>
        public string GetEffectName()
        {
            return _effectName;
        }
        
        /// <summary>
        /// Sets the effect type
        /// </summary>
        public void SetEffect(EffectType effect)
        {
            if (effect == null)
                return;
                
            _effect = effect;
            _effectName = effect.DataName;
        }

        /// <summary>
        /// Starts a new effect on a target
        /// </summary>
        public void StartEffect(EffectType effect, IAttributeObject source, int identifier, int skill = -1)
        {
            _identifier = identifier;
            _skillSource = skill;
            
            if (effect == null)
                return;
                
            SetEffect(effect);
            
            // Start the timer based on effect duration
            _lifeTime = effect.GetDuration(0, source) * 1000;
            
            _attackerId = -1;
            _sourceCastRating = 10.0f;
            
            if (source != null)
            {
                _attackerId = source.GetIdNumber();
                _sourceCastRating = source.GetAbility(Attributes.Ability.CastRating);
            }
        }

        /// <summary>
        /// Updates the effect and applies its ongoing effects to the target
        /// </summary>
        /// <param name="target">Target affected by this effect</param>
        /// <param name="diffTime">Time elapsed in milliseconds</param>
        /// <returns>True if effect should continue, false if it should be removed</returns>
        public bool RunEffect(IAttributeObject target, int diffTime)
        {
            if (target == null || _effect == null)
                return false;
                
            // Decrease timer
            _lifeTime -= diffTime;
            
            if (_lifeTime <= 0)
                return false;
                
            _seconds += diffTime;
            
            // Process effect based on type
            switch (_effect.Type)
            {
                case (int)EffectMechanismType.Poison:
                    if (_seconds >= 4000)  // Poison deals damage every 4 seconds
                    {
                        _seconds -= 4000;
                        target.HarmAttributeByEffect(
                            _effect.GetParameter(0, _sourceCastRating) * 4.0f,
                            0,
                            _attackerId,
                            _effect);
                    }
                    break;
                    
                case (int)EffectMechanismType.Light:
                    // Light effects are passive and don't need periodic updates
                    break;
                    
                case (int)EffectMechanismType.Regenerate:
                    if (_seconds >= 4000)  // Regeneration occurs every 4 seconds
                    {
                        _seconds -= 4000;
                        target.RestoreAttributeByEffect(
                            _effect.GetParameter(1, _sourceCastRating) * 4.0f,
                            (int)_effect.GetParameter(0),
                            _attackerId,
                            _effect);
                    }
                    break;
            }
            
            return true;
        }

        /// <summary>
        /// Gets the source ID of the entity that applied this effect
        /// </summary>
        public int GetSourceID()
        {
            return _attackerId;
        }

        /// <summary>
        /// Gets the remaining life time in milliseconds
        /// </summary>
        public int GetRemainingLifeMillis()
        {
            return _lifeTime;
        }
    }
}